src/compiler/moduleSpecifiers.ts

```diff
@@ -347,7 +347,8 @@ namespace ts.moduleSpecifiers {
         // Get a path that's relative to node_modules or the importing file's path
         // if node_modules folder is in this folder or any of its parent folders, no need to keep it.
         const pathToTopLevelNodeModules = getCanonicalFileName(moduleSpecifier.substring(0, parts.topLevelNodeModulesIndex));
-        if (!(startsWith(sourceDirectory, pathToTopLevelNodeModules) || globalTypingsCacheLocation && startsWith(getCanonicalFileName(globalTypingsCacheLocation), pathToTopLevelNodeModules))) {
+        const canonicalSourceDirectory = getCanonicalFileName(sourceDirectory);
+        if (!(startsWith(canonicalSourceDirectory, pathToTopLevelNodeModules) || globalTypingsCacheLocation && startsWith(getCanonicalFileName(globalTypingsCacheLocation), pathToTopLevelNodeModules))) {

```

Comment:
Wait this seems incorrect. `sourceDirectory` is already canonical path.
https://github.com/microsoft/TypeScript/pull/38462/files#diff-aa86ac6315ba3e1d25a9a1789f7f5a78L110

Reply:
Since this is the only substantive change outside of tests in this PR, is there something we are doing incorrectly when computing `sourceDirectory`? If not, what does this PR change other than the VFS in the test harness and adding a test?

Reply:
I think we would need to diagnose why source directory is not canonical already instead.

Reply:
https://github.com/microsoft/TypeScript/pull/38462#issuecomment-626487827
Would it be better to add comment to let users know it is their responsibility to make sure sourceDirectory canonical path?